### ISSUE TITLE
UID based on format prefix, GPS coordinate and time of issue

### ISSUE THREAD

**Issue by JediJeremy** [openaq/project-universal-stationID#11]
TL;DR: Use a format of `G[latitude]_[longitude]_[datetime]` or `R[random hex number]`

While conceptually simple, any location-based system has to deal with the map datum (co-ordinate system being used), in the same way that any datetime has to deal with the timezone. The simplest way is to fix those choices, for example to WGS84 (also known as EPSG:3857) for the map datum, and GMT (+00:00) for the time. However that has potential long-term consequences because map datums become obsolete over long enough timescales.

On the other hand, you don't want to be so open as to allow any possible datum or timezone, because that puts an onerous requirement on decoding software that may wish to interpret station IDs and arbitrarily lengthens the code.

The simplest way out of that is to add a prefix that indexes into a very short list of legal possibilities, which might also be used to select between this format proposal and others you might like the look of.

For example, you might allocate the prefix code "G" to indicate the GPS based scheme, and "R" for UID's generated by a purely random number generator in cases where no GPS system is available or is inappropriate, and reserve the other letters for future schemes. If the prefix is recognized, the UID can be parsed for further information. If not the ID is still unique, just not as informative.

Latitude and Longitude are probably best represented as decimal degrees, since then co-ordinates can be as accurate as the equipment allows. (five decimal places equates to about 1 meter of precision, more than that is generally superfluous unless you're setting up an array of sensors and wish to represent that information) A choice to be made is whether to allow +/- values for the latitude, or a N/S suffix/separator. Negative values are increasingly standard to define coordinate, so that would preclude its use as a field delimiter.

Since IDs are likely to be translated into web addresses at some point, we also want to avoid the reserved URL/html delimiters like ":&?;=[]()/<>+" and so forth. Decimal values require the "." to be used for that purposes, so precludes its as a field delimiter.

Encoding time also has a range of choices. Unix time is tempting (seconds after midnight 1970) if brevity is more important that readability. Otherwise the standard ISO format of "yyyy-MM-ddThh:mm:ss" looks promising if we substitute the web-unfriendly colons, or use the delimiter omitted option "yyyyMMddThhmmss". Omitting the "-:" separators entirely is probably the best choice, since it maintains compatibility with the standard as well as reducing length, at very little cost in readability. There is an implied "Z" at the end indicating GMT but our code prefix already supplies that information so it's technically redundant.

Similar to the arbitrary-length decimal lat/lon coordinates, we could have a relaxed attitude to precision and allow shorter or longer datetimes without worrying too much about specifying an exact format. The case could be made that an operator is unlikely to commission two unique stations at the exact same location on the same day, so perhaps just "yyyyMMdd" is necessary. But if a device is generating ID's and wishes to ensure absolute uniqueness it can go to the millisecond if it wishes. Or you could specify just the year, if you're setting up a station intended to have that kind of longevity for it's ID.

I'm also a fan of declaring whitespace as "insignificant", so that it may be used to improve readability, but is ignored by any parser by the simple expedient of stripping any tabs/plus/spaces/newlines before comparing IDs.

I'm tempted to recommend using the comma "," as the separator, but I think the underscore is probably the most compatible across systems. 

So for a fixed WGS84 + GMT scheme with the prefix "G", a UID might read:

G-26.13370_-153.60085_20190215

A "long" UID (specified to the millimeter and millisecond) might be:

G-26.13370004_-153.60085003_20190215T0820550900

And if needed it's not too hard to figure out that both ID's are probably referring to the same station, if one somehow got mangled.

So UID's might be anywhere from 30 to 50 characters long depending on precision. Allocating 128 ASCII characters as the maximum possible UID is probably a good idea, to cover all possibilities for this scheme as well as providing extra capacity for future extensions to the syntax.

I'd also define a "random' ID prefix such as "R", followed by at least 32 hexadecimal characters of high-quality randomness. eg:

Rc8425a4a7d9c265db08bec2df21c8f5c

as a fallback for the rare cases when the WGS84+GMT UID cannot be determined. There are web services that will provide these kinds of strings. This scheme may be useful for sensors that are expected to be constantly mobile or may wish "location privacy" - since while there's no conceptual problem allocating the UID based on it's initial position (even for highly mobile sensors such as satellites), it may not always be appropriate.

The advantages of this scheme are a simple method of ensuring uniqueness, and obvious methods of preventing and resolving conflicts - by physically looking around to make sure no-one else is setting up a station within the precision limits of your GPS system at the same time, and then yelling at them if they are. 

If the equipment has an attached GPS receiver, all information (including datetime accurate to the millisecond) is available. If not, most people are walking around with smartphones capable of supplying this information to be entered manually, or have access to a map and a clock. 

A final concern is the allocation of the prefix code. As a general rule, I'd recommend using single-character codes until 12 have been allocated (from the A-Z letters) and then switch to double-character codes until half of the remaining space has been consumed, and so on. That ensures no code is a prefix of any longer code. (so the "G" and "R" codes would never be later used in a double-character code that would confuse earlier software) This prefix list would need to be extended many times over decades for that to become an issue, but it's wise to plan ahead.

Best of luck! 
@JediJeremy
